File: src/main/java/dev/obscuria/tooltips/client/tooltip/filter/PropertyFilter.java
package dev.obscuria.tooltips.client.tooltip.filter;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import net.minecraft.item.ItemStack;
import net.minecraft.util.JsonUtils;

@Desugar
public record PropertyFilter(Boolean hasFoil) implements ItemFilter {

    public static PropertyFilter fromJson(JsonObject json) {
        final Boolean hasFoil = JsonUtils.hasField(json, "has_foil")
                ? JsonUtils.getBoolean(json, "has_foil")
                : null;
        return new PropertyFilter(hasFoil);
    }

    @Override
    public boolean test(ItemStack stack) {
        return hasFoil == null || hasFoil == stack.hasEffect();
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/PropertyFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.world.item.ItemStack
import java.util.*
import kotlin.jvm.optionals.getOrNull

class PropertyFilter(
    val hasFoil: Optional<Boolean>
): IItemFilter {

    override fun codec(): Codec<PropertyFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        return hasFoil.getOrNull()?.equals(stack.hasFoil()) == true
    }

    companion object {

        val CODEC: Codec<PropertyFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                Codec.BOOL
                    .optionalFieldOf("has_foil")
                    .forGetter(PropertyFilter::hasFoil)
            ).apply(codec, ::PropertyFilter)
        }
    }
}